drivers/mge-hid.c

```diff
@@ -300,28 +396,99 @@ static info_lkp_t eaton_abm_status_info[] = {
 	{ 0, NULL, NULL, NULL }
 };
 
+/* Used to process ABM flag, for battery.charger.type */
+static const char *eaton_abm_charger_type_fun(double value)
+{
+	int abm_charger_type = value;
+
+	if (abm_charger_type == ABM_ENABLED_TYPE)
+	{
+		/* Set ABM flag for battery.charger.type */
+		advanced_battery_monitoring = ABM_ENABLED_TYPE;
+		upsdebugx(2, "ABM charger type: %i", (int)value);
+
+		return "ABM";
+	}
+
+    /* Handle the case when value is not equal to ABM_ENABLED_TYPE_MODE or some other appropriate action */
+	return NULL;
+}
+
+static info_lkp_t eaton_charger_type_info[] = {
+	{ 0, "None", NULL, NULL },
+	{ 1, "Extended (CLA)", NULL, NULL },
+	{ 2, "Large extension", NULL, NULL },
+	{ 3, "Extra large extension (XL)", NULL, NULL },
+	{ 4, "ABM", eaton_abm_charger_type_fun, NULL },
+	{ 5, "Constant Charge (CC)", NULL, NULL },
+	{ 0, NULL, NULL, NULL }
+};
+
 /* Used to process ABM flags, for ups.status (CHRG/DISCHRG/RB) */
 static const char *eaton_abm_chrg_dischrg_fun(double value)
 {
 	/* Don't process if ABM is disabled */
 	if (advanced_battery_monitoring == ABM_DISABLED)
 		return NULL;
 
-	switch ((long)value)
+	/* if we have "x.Status" for 9E Models and others but not "x.Mode" */
+	if ((advanced_battery_status == ABM_CHARGER_STATUS) && (advanced_battery_mode != ABM_CHARGER_MODE))
 	{
-	case 1: /* charging status */
-	case 3: /* floating status */
-		snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "chrg");
-		break;
-	case 2:
-		snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "dischrg");
-		break;
-	case 6: /* ABM Charger Disabled */
-	case 4: /* resting, nothing to publish! (?) */
-	case 5: /* Undefined - ABM is not activated */
-	default:
-		/* Return NULL, not to get the value published! */
-		return NULL;
+		switch ((long)value)
+		{
+		case 1: /* charging status, FIXME: 9E Model - no chrg in ups.status when battery.charge < 100 */
+			snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "chrg");
+			break;
+
+		case 2: /* floating status, FIXME: 9E Model - chrg not stop in ups.status after battery.charge = 100 */
+			if (advanced_battery_monitoring == ABM_UNKNOWN)
+			{
+				snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "!chrg");
+				break;
+			}
+			else
+			{
+				/* charging status, floating status */
+				snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "chrg");
+				break;
+			}
+		case 4: /* discharging status, FIXME: 9E Model - remove dischrg in ups.status when on battery */
+			if (advanced_battery_monitoring == ABM_UNKNOWN)
+			{
+				snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "!dischrg");
+				break;
+			}
+			else
+			{
+				snprintf(mge_scratch_buf, sizeof(mge_scratch_buf), "%s", "dischrg");
+				break;
+			}
```

Comment:
Not sure about these `if (advanced_battery_monitoring == ABM_UNKNOWN)`, why are we returning that we are not charging or discharging here? Would it not be better logically to either assume ABM is on (as ABM paths are present and ABM is not overall `ABM_DISABLED`) or exit the function prematurely (as is done for `ABM_DISABLED` earlier on L307/431)?

General charging and discharging is separately handled in `eaton_abm_check_chrg_fun` so - I think - probably we shouldn't clear flags based on `ABM_UNKNOWN` here and just let this function either still run or not run when ABM status is still `ABM_UNKNOWN`.

I personally think better would be to not let this function run when ABM is not specifically known to be enabled (yet), especially if we perhaps also don't know which ABM path it will take (yet), so I'd treat it the same as `ABM_DISABLED` and `return NULL` as on L307/431. At least from my current understanding...